File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/logic/gametasks/hud/elements/player/WrongWay.java
package com.bitfire.uracer.game.logic.gametasks.hud.elements.player;

import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.SpriteBatch;
import com.bitfire.uracer.configuration.Config;
import com.bitfire.uracer.game.logic.gametasks.hud.Positionable;
import com.bitfire.uracer.game.tween.GameTweener;
import com.bitfire.uracer.resources.Art;
import com.bitfire.uracer.utils.AMath;
import com.bitfire.uracer.utils.BoxedFloat;
import com.bitfire.uracer.utils.BoxedFloatAccessor;

import aurelienribon.tweenengine.Timeline;
import aurelienribon.tweenengine.Tween;
import aurelienribon.tweenengine.equations.Linear;

public class WrongWay extends Positionable {
    private final float w;
    private final float h;
    private final float offX;
    private final float offY;
    private final Sprite sign;
    private final BoxedFloat bfAlpha;
    private boolean isShown;

    public WrongWay() {
        sign = new Sprite(Art.wrongWay);
        float scale = 0.4f;
        w = Art.wrongWay.getWidth() * scale;
        h = Art.wrongWay.getHeight() * scale;
        offX = w / 2;
        offY = h / 2;

        sign.setSize(w, h);
        sign.setOrigin(offX, offY);
        sign.flip(false, true);
        bfAlpha = new BoxedFloat(0);
        isShown = false;
    }

    @Override
    public float getWidth() {
        return w;
    }

    @Override
    public float getHeight() {
        return h;
    }

    public void fadeIn() {
        fadeIn(Config.Graphics.DefaultFadeMilliseconds);
    }

    public void fadeIn(int millisecs) {
        if (!isShown) {
            isShown = true;
            GameTweener.stop(bfAlpha);
            Timeline seq = Timeline.createSequence();
            seq.push(Tween.to(bfAlpha, BoxedFloatAccessor.VALUE, millisecs).target(1f).ease(Linear.INOUT));
            GameTweener.start(seq);
        }
    }

    public void fadeOut(int millisecs) {
        if (isShown) {
            isShown = false;
            GameTweener.stop(bfAlpha);
            Timeline seq = Timeline.createSequence();
            seq.push(Tween.to(bfAlpha, BoxedFloatAccessor.VALUE, millisecs).target(0f).ease(Linear.INOUT));
            GameTweener.start(seq);
        }
    }

    public void render(SpriteBatch batch) {
        if (!AMath.isZero(bfAlpha.value)) {

            float px = position.x - offX;
            float py = position.y - offY;

            sign.setPosition(px, py);

            sign.draw(batch, bfAlpha.value);
        }
    }
}

File: DriftGame/uracer-desktop/src/com/bitfire/uracer/game/logic/gametasks/hud/elements/player/WrongWay.kt
package com.bitfire.uracer.game.logic.gametasks.hud.elements.player

import aurelienribon.tweenengine.Timeline
import aurelienribon.tweenengine.Tween
import aurelienribon.tweenengine.equations.Linear.Companion.INOUT
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.graphics.g2d.SpriteBatch
import com.bitfire.uracer.configuration.GraphicsUtils
import com.bitfire.uracer.game.logic.gametasks.hud.Positionable
import com.bitfire.uracer.game.tween.GameTweener.start
import com.bitfire.uracer.game.tween.GameTweener.stop
import com.bitfire.uracer.resources.Art
import com.bitfire.uracer.utils.AlgebraMath.isZero
import com.bitfire.uracer.utils.BoxedFloat
import com.bitfire.uracer.utils.BoxedFloatAccessor

class WrongWay : Positionable() {

    private val w: Float
    private val h: Float
    private val offX: Float
    private val offY: Float
    private val sign = Sprite(Art.wrongWay)
    private val bfAlpha: BoxedFloat
    private var isShown: Boolean

    init {
        val scale = 0.4F
        w = Art.wrongWay.width * scale
        h = Art.wrongWay.height * scale
        offX = w / 2
        offY = h / 2

        sign.setSize(w, h)
        sign.setOrigin(offX, offY)
        sign.flip(false, true)
        bfAlpha = BoxedFloat(0F)
        isShown = false
    }

    override val width = w
    override val height = h

    @JvmOverloads
    fun fadeIn(millisecs: Int = GraphicsUtils.DefaultFadeMilliseconds) {
        if (!isShown) {
            isShown = true
            stop(bfAlpha)
            val seq = Timeline.createSequence()
            seq.push(Tween.to(bfAlpha, BoxedFloatAccessor.VALUE, millisecs.toFloat()).target(1f).ease(INOUT))
            start(seq)
        }
    }

    fun fadeOut(millisecs: Int) {
        if (isShown) {
            isShown = false
            stop(bfAlpha)
            val seq = Timeline.createSequence()
            seq.push(Tween.to(bfAlpha, BoxedFloatAccessor.VALUE, millisecs.toFloat()).target(0F).ease(INOUT))
            start(seq)
        }
    }

    fun render(batch: SpriteBatch) {
        if (!isZero(bfAlpha.value)) {
            val px = position.x - offX
            val py = position.y - offY
            sign.setPosition(px, py)
            sign.draw(batch, bfAlpha.value)
        }
    }
}
